File: app/src/main/java/com/elementary/tasks/core/data/dao/ReminderDao.java
package com.elementary.tasks.core.data.dao;

import com.elementary.tasks.core.data.models.Reminder;

import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;

import static androidx.room.OnConflictStrategy.REPLACE;

/**
 * Copyright 2018 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@Dao
public interface ReminderDao {

    @Query("SELECT * FROM Reminder WHERE uniqueId=:id")
    LiveData<Reminder> loadById(int id);

    @Query("SELECT * FROM Reminder WHERE noteId=:key")
    LiveData<Reminder> loadByNoteKey(String key);

    @Query("SELECT * FROM Reminder WHERE uniqueId=:id")
    Reminder getById(int id);

    @Query("SELECT * FROM Reminder WHERE isActive=:active AND isRemoved=:removed")
    LiveData<List<Reminder>> loadType(boolean active, boolean removed);

    @Query("SELECT * FROM Reminder WHERE isRemoved=:removed AND eventTime!=0 AND eventTime>=:fromTime AND eventTime<:toTime")
    List<Reminder> getActiveInRange(boolean removed, String fromTime, String toTime);

    @Query("SELECT * FROM Reminder WHERE isActive=:active AND isRemoved=:removed")
    List<Reminder> getAll(boolean active, boolean removed);

    @Query("SELECT * FROM Reminder WHERE isActive=:active AND isRemoved=:removed AND type IN (:types)")
    List<Reminder> getAllTypes(boolean active, boolean removed, int[] types);

    @Query("SELECT * FROM Reminder WHERE isActive=:active AND isRemoved=:removed AND type IN (:types)")
    LiveData<List<Reminder>> loadAllTypes(boolean active, boolean removed, int[] types);

    @Query("SELECT * FROM Reminder WHERE isActive=:active AND isRemoved=:removed AND eventTime!=0 AND eventTime>=:fromTime AND eventTime<:toTime")
    List<Reminder> getAllTypesInRange(boolean active, boolean removed, String fromTime, String toTime);

    @Insert(onConflict = REPLACE)
    int insert(Reminder reminder);

    @Insert(onConflict = REPLACE)
    void insertAll(Reminder... reminder);

    @Delete
    void delete(Reminder reminder);

    @Delete
    void deleteAll(Reminder... reminder);
}

File: app/src/main/java/com/elementary/tasks/core/data/dao/ReminderDao.kt
package com.elementary.tasks.core.data.dao

import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction
import com.elementary.tasks.core.data.models.Reminder

private const val byIdQuery = """
    SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
    FROM Reminder AS reminder
    JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
    WHERE reminder.uuId=:id"""

private const val byNoteIdQuery = """
    SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
    FROM Reminder AS reminder
    JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
    WHERE reminder.noteId=:key"""

@Dao
interface ReminderDao {

  @Query("SELECT * FROM Reminder")
  fun getAll(): List<Reminder>

  @Transaction
  @Query(byIdQuery)
  fun loadById(id: String): LiveData<Reminder?>

  @Transaction
  @Query(byIdQuery)
  fun getById(id: String): Reminder?

  @Transaction
  @Query(byNoteIdQuery)
  fun loadByNoteKey(key: String): LiveData<Reminder>

  @Transaction
  @Query(byNoteIdQuery)
  fun getByNoteKey(key: String): Reminder?

  @Transaction
  @Query(
    """SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
        FROM Reminder AS reminder
        JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
        WHERE reminder.isRemoved=:removed
        AND LOWER(Reminder.summary) LIKE '%' || :query || '%'
        ORDER BY reminder.isActive DESC, reminder.eventTime ASC"""
  )
  fun searchBySummaryAndRemovedStatus(query: String, removed: Boolean = false): List<Reminder>

  @Transaction
  @Query(
    """SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
        FROM Reminder AS reminder
        JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
        WHERE reminder.isRemoved=:removed
        ORDER BY reminder.isActive DESC, reminder.eventTime ASC"""
  )
  fun getByRemovedStatus(removed: Boolean = false): List<Reminder>

  @Transaction
  @Query(
    """SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
        FROM Reminder AS reminder
        JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
        WHERE reminder.isRemoved=:removed
        AND reminder.isActive=:active"""
  )
  fun loadType(active: Boolean, removed: Boolean): LiveData<List<Reminder>>

  @Transaction
  @Query(
    """SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
        FROM Reminder AS reminder
        JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
        WHERE reminder.isRemoved=:removed
        AND reminder.eventTime!=''
        AND reminder.eventTime>=:fromTime
        AND reminder.eventTime<:toTime"""
  )
  fun getActiveInRange(removed: Boolean, fromTime: String, toTime: String): List<Reminder>

  @Transaction
  @Query(
    """SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
        FROM Reminder AS reminder
        JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
        WHERE reminder.isRemoved=:removed
        AND reminder.isActive=:active
        AND reminder.eventTime!=''
        AND reminder.eventTime>=:fromTime
        AND reminder.eventTime<:toTime"""
  )
  fun getAllTypesInRange(
    active: Boolean,
    removed: Boolean,
    fromTime: String,
    toTime: String
  ): List<Reminder>

  @Transaction
  @Query(
    """SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
        FROM Reminder AS reminder
        JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
        WHERE reminder.isRemoved=:removed
        AND reminder.isActive=:active"""
  )
  fun getAll(active: Boolean, removed: Boolean): List<Reminder>

  @Transaction
  @Query(
    """SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
        FROM Reminder AS reminder
        JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
        WHERE reminder.isRemoved=:removed
        AND reminder.isActive=:active
        AND reminder.type IN (:types)"""
  )
  fun getAllTypes(active: Boolean, removed: Boolean, types: IntArray): List<Reminder>

  @Transaction
  @Query(
    """SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
        FROM Reminder AS reminder
        JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
        WHERE reminder.isRemoved=:removed
        AND reminder.isActive=:active
        AND reminder.type IN (:types)"""
  )
  fun loadAllTypes(active: Boolean, removed: Boolean, types: IntArray): LiveData<List<Reminder>>

  @Transaction
  @Query(
    """SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
        FROM Reminder AS reminder
        JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
        WHERE reminder.isRemoved=:removed
        AND reminder.isActive=:active
        AND reminder.eventTime!=''
        AND reminder.eventTime>=:fromTime
        AND reminder.eventTime<:toTime
        ORDER BY reminder.eventTime ASC"""
  )
  fun loadAllTypesInRange(
    active: Boolean = true,
    removed: Boolean = false,
    fromTime: String,
    toTime: String
  ): LiveData<List<Reminder>>

  @Transaction
  @Query(
    """SELECT reminder.*, g.groupTitle, g.groupUuId, g.groupColor
        FROM Reminder AS reminder
        JOIN ReminderGroup AS g ON reminder.groupUuId = g.groupUuId
        WHERE (reminder.isActive=:active AND reminder.isRemoved=:removed)
        AND (reminder.eventTime=='' OR (reminder.eventTime>=:fromTime AND reminder.eventTime<:toTime))
        ORDER BY reminder.eventTime ASC"""
  )
  fun loadAllTypesInRangeWithPermanent(
    active: Boolean = true,
    removed: Boolean = false,
    fromTime: String,
    toTime: String
  ): LiveData<List<Reminder>>

  @Insert(onConflict = OnConflictStrategy.REPLACE)
  fun insert(reminder: Reminder)

  @Insert(onConflict = OnConflictStrategy.REPLACE)
  fun insertAll(reminders: Iterable<Reminder>)

  @Delete
  fun delete(reminder: Reminder)

  @Query("DELETE FROM Reminder WHERE uuId=:id")
  fun delete(id: String)

  @Delete
  fun deleteAll(reminders: Iterable<Reminder>)

  @Query("DELETE FROM Reminder")
  fun deleteAll()
}
